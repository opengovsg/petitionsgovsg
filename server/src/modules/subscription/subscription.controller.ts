import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { CreateSubscriptionReqDto, ErrorDto } from '~shared/types/api'
import { createLogger } from '@/bootstrap/logging'
import { ControllerHandler } from '@/types/response-handler'
import { SubscriptionService } from './subscription.service'

const logger = createLogger(module)

export class SubscriptionController {
  private subscriptionService: Public<SubscriptionService>

  constructor({
    subscriptionService,
  }: {
    subscriptionService: Public<SubscriptionService>
  }) {
    this.subscriptionService = subscriptionService
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @body email
   * @returns 200 with new subscription id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 403 if user is not authorized to answer question
   * @returns 500 if database error
   */
  createSubscription: ControllerHandler<
    { id: string },
    number | ErrorDto,
    CreateSubscriptionReqDto,
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: errors.array()[0].msg })
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      // Save Subscription in the database
      const subscriptionId = await this.subscriptionService.createSubscription({
        email: req.body.email,
        postId: Number(req.params.id),
      })
      return res.status(StatusCodes.OK).json(subscriptionId)
    } catch (error) {
      logger.error({
        message: 'Error while adding new subscription',
        meta: {
          function: 'createSubscription',
          userId: req.user.id,
          postId: req.params.id,
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }
  }
}
