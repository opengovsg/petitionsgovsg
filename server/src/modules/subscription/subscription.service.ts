import type { Sequelize as SequelizeType } from 'sequelize'
import { Subscription } from '~shared/types/base'
import { ModelDef } from '@/types/sequelize'

export class SubscriptionService {
  private Subscription: ModelDef<Subscription>
  private sequelize: SequelizeType

  constructor({
    Subscription,
    sequelize,
  }: {
    Subscription: ModelDef<Subscription>
    sequelize: SequelizeType
  }) {
    this.Subscription = Subscription
    this.sequelize = sequelize
  }

  /**
   * Create a subscription attached to a post
   * @param postId id of post to attach to
   * @param email subscriber's email
   * @returns id of new subscription if it is successfully created
   */
  createSubscription = async ({
    postId,
    email,
  }: Pick<Subscription, 'email' | 'postId'>): Promise<number> => {
    const subscriptionId = await this.sequelize.transaction(
      async (transaction) => {
        const subscription = await this.Subscription.create(
          {
            postId: postId,
            email: email,
          },
          { transaction },
        )
        return subscription.id
      },
    )
    return subscriptionId
  }
}
