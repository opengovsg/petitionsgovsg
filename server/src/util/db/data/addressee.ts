export const mockAddressee = {
  name: 'Ministry of Environment',
  shortName: 'MoE',
}

export const mockAddressees = [
  {
    id: 1,
    name: 'Ministry of Environment',
    shortName: 'MoE',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
  {
    id: 2,
    name: 'Ministry of Labour',
    shortName: 'MoL',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
    openToPublic: false,
    officerEmails: [],
  },
]
