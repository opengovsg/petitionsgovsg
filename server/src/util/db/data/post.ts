import { PostStatus } from '~shared/types/base'

export const mockPost = {
  id: '395ac6ec-dea2-4f78-8dc4-1234cafaff87',
  title: 'Basic petition without optional items',
  reason: 'reason for',
  request: 'request',
  status: PostStatus.Open,
  fullname: 'John Doe',
  salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
  hashedUserSgid: 'wxyz',
  email: 'John@email.com',
  addresseeId: 1,
  summary: null,
  references: null,
  profile: null,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockPosts = [
  {
    id: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    title: 'Basic petition without optional items',
    reason: 'reason for',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
    hashedUserSgid:
      '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
    email: 'John@email.com',
    addresseeId: 1,
    summary: null,
    references: null,
    profile: null,
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 'aaaaaaaa-dea2-4f78-8dc4-5542cafaff87',
    title: 'Petition with summary, references, profile',
    reason: 'reason',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tX.',
    hashedUserSgid: 'wxyz',
    email: 'John@email.com',
    addresseeId: 1,
    summary: 'petition summary',
    references: 'wikipedia.com',
    profile: 'OGP',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
