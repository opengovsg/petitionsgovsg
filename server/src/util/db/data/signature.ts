export const mockSignature = {
  id: 4,
  hashedUserSgid:
    '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
  fullname: null,
  comment: 'I want to support this cause.',
  postId: 1,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockSignatures = [
  {
    id: 1,
    hashedUserSgid: '1234',
    fullname: null,
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 2,
    hashedUserSgid: '5678',
    fullname: 'John Doe',
    comment: null,
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 3,
    hashedUserSgid: '8765',
    fullname: null,
    comment: 'I support this cause!',
    postId: '1',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
