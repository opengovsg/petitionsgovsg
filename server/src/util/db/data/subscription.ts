export const mockSubscription = {
  id: 1,
  email: 'test@email.com',
  postId: 1,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}
