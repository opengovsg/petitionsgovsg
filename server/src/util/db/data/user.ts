import { UserAuthType } from '~shared/types/api'

export const mockUser = {
  id: 'sgid',
  fullname: 'John Doe',
  type: UserAuthType.Public,
}
