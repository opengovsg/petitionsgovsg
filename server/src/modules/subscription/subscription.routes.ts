import { AuthMiddleware } from '../auth/auth.middleware'
import express from 'express'
import { body } from 'express-validator'
import { SubscriptionController } from './subscription.controller'
import { limiter } from '@/middleware/limiter'

export const routeSubscriptions = ({
  controller,
  authMiddleware,
}: {
  controller: SubscriptionController
  authMiddleware: AuthMiddleware
}): express.Router => {
  const router = express.Router()
  const { authenticate } = authMiddleware

  /**
   * Create a subscription attached to a post
   * @route   POST /api/subscriptions/:id
   * @returns 200 with new subscriptions id
   * @returns 400 if invalid request
   * @returns 401 if user not signed in
   * @returns 500 if database error
   * @access  Private
   */
  router.post(
    '/:id',
    [
      authenticate,
      limiter,
      body('email', 'email should be a valid format').isString(),
    ],
    controller.createSubscription,
  )
  return router
}
