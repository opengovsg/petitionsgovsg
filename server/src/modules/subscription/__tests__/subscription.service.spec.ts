import {
  createTestDatabase,
  resetAndSetupDb,
  SequelizeWithModels,
} from '@/util/db/jest-db'
import { SubscriptionService } from '../subscription.service'
import { mockSubscription } from '@/util/db/data/subscription'

describe('SubscriptionService', () => {
  let service: SubscriptionService
  let db: SequelizeWithModels

  beforeAll(async () => {
    db = await createTestDatabase()
    service = new SubscriptionService({
      Subscription: db.Subscription,
      sequelize: db.sequelize,
    })
  })

  beforeEach(async () => {
    await resetAndSetupDb(db)
    jest.resetAllMocks()
  })

  describe('createSuscription', () => {
    it('creates subscription for specified post id', async () => {
      const subscriptionId = await service.createSubscription(mockSubscription)
      const subscription = await db.Subscription.findByPk(subscriptionId)

      expect(subscription?.get()).toMatchSnapshot({
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      })
    })
  })
})
