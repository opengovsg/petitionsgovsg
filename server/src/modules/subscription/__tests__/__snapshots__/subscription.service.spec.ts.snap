// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionService createSuscription creates subscription for specified post id 1`] = `
Object {
  "createdAt": Any<Date>,
  "email": "test@email.com",
  "id": 1,
  "postId": 1,
  "updatedAt": Any<Date>,
}
`;
